crates/rethnet_evm/benches/state.rs

```diff
@@ -0,0 +1,131 @@
+use std::{clone::Clone, str::FromStr};
+
+use rethnet_eth::Address;
+use rethnet_evm::state::{HybridState, LayeredState, RethnetLayer, StateError, SyncState};
+
+#[derive(Default)]
+struct RethnetStates {
+    layered: LayeredState<RethnetLayer>,
+    hybrid: HybridState<RethnetLayer>,
+}
+
+impl RethnetStates {
+    fn fill(&mut self, number_of_accounts: u64, number_of_layers_per_account: u64) {
+        let mut states: [&mut dyn SyncState<StateError>; 2] = [&mut self.layered, &mut self.hybrid];
+        for state in states.iter_mut() {
+            for i in 1..=number_of_accounts {
+                state
+                    .insert_account(
+                        Address::from_str(&format!("0x{:0>40x}", i)).unwrap(),
+                        AccountInfo::default(),
+                    )
+                    .unwrap();
+                for _ in 1..=number_of_layers_per_account {
+                    state.checkpoint().unwrap();
+                }
+            }
+        }
+    }
+
+    /// Returns a set of factories, each member of which produces a clone of one of the state objects in this struct.
+    fn make_clone_factories(
+        &self,
+    ) -> Vec<(
+        &'static str, // label of the type of state produced by this factory
+        Box<dyn Fn() -> Box<dyn SyncState<StateError>> + '_>,
+    )> {
+        vec![
+            ("LayeredState", Box::new(|| Box::new(self.layered.clone()))),
+            ("HybridState", Box::new(|| Box::new(self.hybrid.clone()))),
+        ]
+    }
+}
+
+const ADDRESS_SCALES: [u64; 4] = [100, 500, 1000, 2000];
+const LAYER_SCALES: [u64; 4] = [1, 2, 4, 8];
+
+fn bench_insert_account(c: &mut Criterion) {
+    for number_of_accounts in ADDRESS_SCALES.iter() {
+        for layers_per_account in LAYER_SCALES.iter() {
+            let mut rethnet_states = RethnetStates::default();
+            rethnet_states.fill(*number_of_accounts, *layers_per_account);
+
+            for (label, state_factory) in rethnet_states.make_clone_factories().into_iter() {
+                c.benchmark_group(format!(
+                    "StateRef::insert_account() with {} accounts with {} layer(s) per account",
+                    *number_of_accounts, layers_per_account,
+                ))
+                .bench_function(label, |b| {
+                    b.iter_batched(
+                        || state_factory(),
+                        |mut state| {
+                            state.insert_account(
+                                Address::from_str(&format!("0x{:0>40x}", number_of_accounts + 1))
+                                    .unwrap(),
+                                AccountInfo::default(),
+                            )
+                        },
+                        BatchSize::SmallInput,
+                    );
+                });
+            }
+        }
+    }
+}
+
+fn bench_checkpoint(c: &mut Criterion) {
+    for number_of_accounts in ADDRESS_SCALES.iter() {
```

Comment:
A lot of this code repeats itself for every benchmark. Could you refactor it to remove code duplication?

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/3868/commits/b1f392822bb51b937ce7008af7b6d1c5f65c019d

I wanted to take it even further, by having a single `c.benchmark_group(...).bench_function(...)`, within a loop over an array of `(group_label, state_action)` tuples, where a `state_action` conveys the second parameter to `iter_batched()`, but I started running into Boxes and return type differences and couldn't figure it out in a reasonable amount of time. If you'd like to see that happen, maybe we can pair on it to expedite things.

Reply:
I don't really understand this change. We should still have separate benchmarks, but they should be reusing the parts of the code that are shared, split off into a separate function.

I'm happy to pair on it tomorrow.

Reply:
Reverted the previous change in https://github.com/NomicFoundation/hardhat/pull/3868/commits/102e7cb25cdc8cce1363be283e962727ee61ba5a and re-did it in https://github.com/NomicFoundation/hardhat/pull/3868/commits/26fcefb1976bcbb561749ea0bfed60f93500154b